vsintegration/src/FSharp.Editor/CodeLens/LineLens.fs

```diff
@@ -0,0 +1,697 @@
+    let bufferChangedEvent = new Event<EventHandler<TextContentChangedEventArgs>,TextContentChangedEventArgs>()
+
+    // Tracks the created ui elements per TrackingSpan
+    let uiElements = Dictionary<_,_>()
+    /// Caches the current used trackingSpans per line. One line can contain multiple trackingSpans
+    let mutable trackingSpans = Dictionary<_, Generic.List<_>>()
+    /// Text view for accessing the adornment layer.
+    let mutable view: IWpfTextView = view
+    /// The line lens layer for adding and removing adornments.
+    let mutable lineLensLayer: IAdornmentLayer = view.GetAdornmentLayer "LineLens"
+    /// Tracks the recent first + last visible line numbers for adornment layout logic.
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    /// Tracks the adornments on the layer.
+    let mutable addedAdornments = HashSet()
+    /// Cancellation token source for the layout changed event. Needed to abort previous async-work.
+    let mutable layoutChangedCts = new CancellationTokenSource()
+
+    /// Tracks the last used buffer snapshot, should be preferred used in combination with mutex.
+    let mutable currentBufferSnapshot = null
+
+    /// Helper method which returns the start line number of a tracking span
+    let getTrackingSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        snapshot.GetLineNumberFromPosition(trackingSpan.GetStartPoint(snapshot).Position)
+
+    /// Helper method which returns the start line number of a tracking span
+    let tryGetTSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        let pos = trackingSpan.GetStartPoint(snapshot).Position
+        if snapshot.Length - 1 < pos then None
+        else pos |> snapshot.GetLineNumberFromPosition |> Some
+    
+    let updateTrackingSpansFast (snapshot:ITextSnapshot) lineNumber =
+        if lineNumber |> trackingSpans.ContainsKey then
+            let currentTrackingSpans = Generic.List(trackingSpans.[lineNumber])
+            for trackingSpan in currentTrackingSpans do
+                let newLineOption = tryGetTSpanStartLine snapshot trackingSpan
+                match newLineOption with 
+                | None -> ()
+                | Some newLine ->
+                    if newLine <> lineNumber then
+                        trackingSpans.[lineNumber].Remove(trackingSpan) |> ignore
+                        if trackingSpans.[lineNumber].Count = 0 then
+                            trackingSpans.Remove lineNumber |> ignore
+                        if newLine |> trackingSpans.ContainsKey |> not then
+                            trackingSpans.[newLine] <- Generic.List()
+                        trackingSpans.[newLine].Add(trackingSpan)
+                        // tagsChangedEvent.Trigger(self, SnapshotSpanEventArgs(trackingSpan.GetSpan snapshot)) // This results in super annoying blinking
+
+    let createDefaultStackPanel () = 
+        let uiElement = new StackPanel()
```

Comment:
I'd recommend using a Grid or other container (or no container if possible) here - If use Grid, and can set it's height correctly, that'd allow the textblock to use VerticalAlignment="Center", and fix your vertical alignment issues.

Reply:
I think I must do this in addition to adjust the manual layout not overlaying two elements.  

To reproduce just have more than one constant or function declarations on one line.